lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -320,6 +321,128 @@ void getAsmBlockArgumentNamesImpl(Operation *op, mlir::Region &region,
 static ParseResult parseNameKind(OpAsmParser &parser,
                                  firrtl::NameKindEnumAttr &result);
 
+//===----------------------------------------------------------------------===//
+// Layer Verification Utilities
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct CompareSymbolRefAttr {
+  // True if lhs is lexicographically less than rhs.
+  bool operator()(SymbolRefAttr lhs, SymbolRefAttr rhs) const {
+    auto cmp = lhs.getRootReference().compare(rhs.getRootReference());
+    if (cmp == -1)
+      return true;
+    if (cmp == 1)
+      return false;
+    auto lhsNested = lhs.getNestedReferences();
+    auto rhsNested = rhs.getNestedReferences();
+    auto lhsNestedSize = lhsNested.size();
+    auto rhsNestedSize = rhsNested.size();
+    auto e = std::min(lhsNestedSize, rhsNestedSize);
+    for (unsigned i = 0; i < e; ++i) {
+      auto cmp = lhsNested[i].getAttr().compare(rhsNested[i].getAttr());
+      if (cmp == -1)
+        return true;
+      if (cmp == 1)
+        return false;
+    }
+    return lhsNestedSize < rhsNestedSize;
+  }
+};
+} // namespace
+
+using LayerSet = SmallSet<SymbolRefAttr, 4, CompareSymbolRefAttr>;
```

Comment:
> Specifically, are the first three redundant informationally? I would think yes.

Yes!

> This does hit the "fast lookup path" in isLayerCompatible but only when stored in such a fashion (which is not guaranteed given how enablelayer is pulled in with, e.g., Y.Z, and will hit the "slow path").

And how the explicit layer requirements of a probe type are pulled in.

> This may motivate a slightly more complex data structure that avoids redundant storage while still making lookup fast.

Yep!

There are two things I like about how I've implemented it:

1) when we tell the user what layers are missing, we only tell them the missing layers that were actually written in the source. So if a missing layer requirement comes from nested layerblocks A and B, we'll list both A and A.B, if it comes from a probe color like A.B, we only list A.B.

2) I wrote this code so that it would be easier to make changes later on, in particular I kept an eye on what it would take to remove nested layer declarations.